design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,259 @@
+* `VirtualChunkContainers` serve several purposes:
+  * Using their `name` a user can setup credentials for a specific set of chunks, in a given object store and bucket prefix.
+  * Using their `protocol` and other properties, Icechunk knows how to build an object store client that can retrieve the virtual chunks.
+  * Users can review the list of `VirtualChunkContainers` for a repo to quickly understand what other buckets and prefixes contain data used by this repo.
+  * Users can review the list of `VirtualChunkContainers` for a repo to quickly understand what credentials they need to provide if they want to access all virtual chunks. They can still use the repo without providing those credentials, but they virtual chunk requests may fail.
+
+### Global config
+
+The list of chunk containers is stored in a new global `Config` object. There is a single configuration per repo, and all operations use the last version of the configuration.
+
+The configuration is stored as a JSON file in `repo-root/config.json`, and it's written using PUT if-match to ensure writers are aware of the latest updates.
+
+The global configuration solves also [other requirements](https://github.com/earth-mover/icechunk/issues/440) outside of this design doc. Icechunk repos need, also for other reasons, a way to store persistent default configuration.
+
+```rust
+// this is pseudo-code
+struct Config {
+  inline_chunk_threshold_bytes: u16,
+  ... more fields
+
+  virtual_chunk_containers: Vec<VirtualChunkContainer>,
+}
+```
+
+The `Config` object acts as a default, overrideable by users at runtime. But different fields of configuration may have different behaviors on overrides.
+
+### New mechanism to set virtual references
+
+We add an optional `checksum` argument to `set_virtual_ref` (we need a better name for this argument).
+
+```python
+# this is pseudo-code
+
+class ETag:
+  value: str
+
+class LastModified:
+  value: datetime
+
+class Repository:  # the class with this method is going to change soon to be Session, for an unrelated reason
+  def set_virtual_ref(
+      self,
+      key: str,
+      location: str,
+      offset: int,
+      length: int,
+      checksum: ETag | LastModified | None = None,
+      validate_containers: bool = False
+  ) -> None: ...
+```
+
+`checksum` can be set to the in-object-store ETag of the file containing the virtual chunk, or to its last modified time. In the case of last modified time, the value
+doesn't need to be precise. The system will only fail to provide the chunk if the object's last-modified-at is larger than the passed value. So, passing `checksum = LastModified(datetime.now())` will usually be enough.
+
+This new argument will help us satisfy requirement 4.
+
+The interpretation of the `location` argument also changes. `location` must be an absolute url that includes a protocol. The protocol can be:
+
+* s3://
+* gcs://
+* tigris://  (is this the right protocol)
```

Comment:
No. it's just S3 with a custom endpoint URL.

Reply:
There is also credentials... We may need to know it's tigris so we manage the generation of credentials differently from say localhost minio. Of course, we can always play games with the endpoint and figure out, but this seems more future-proof?

Reply:
> we manage the generation of credentials differently from say localhost minio

I recommend you keep to using ("s3") for everything that indeed matches that protocol, and the rest is in the configuration for the individual "containers" - isn't that the whole point? I don't think you want to maintain a potentially endless list of protocols for each and service that might have some icy data.

> generation of credentials

What does this mean?

By the way, how exactly are chunk URLs mapped to "containers", i.e., how do you know which container a particular URL belongs to?

Reply:
One downside of using `s3://` everywhere is that it forces users to actually create the VirtualChunkContainers, even for some very common object stores like Google's. Using the protocol would be a bit easier on them for the very common case of a simple repo on google or tigris. But I take your point, which ones are s3:// and which ones are a protocol is arbitrary. Will think more about it.

Reply:
>  how exactly are chunk URLs mapped to "containers", 

@martindurant a rough description of the algorithm is documented [here](https://github.com/earth-mover/icechunk/pull/436/files#diff-c698aed6bffbd27180af5e8495caefc5b69f58680485b27b1dbbbdbb0788bcf0R141)

Reply:
> One downside of using s3:// everywhere is that it forces users to actually create the VirtualChunkContainers, even for some very common object stores like Google's. 

?? google's storage does not use the s3 protocol. When I say "protocol", mean exactly that: a set of rules and conventions for how to communicate. S3's is [defined by AWS](https://docs.aws.amazon.com/AmazonS3/latest/API/API_Operations_Amazon_Simple_Storage_Service.html) but implemented by many services. GCS is *only* used by google (as far as I know).

> a rough description of the algorithm

made some additional comments